### ISSUE TITLE
JavadocMethod doesn't report anonymous class methods

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#2827]
TestClass.java

```
package test;

public class TestClass {
    public static final MyClass myField = new MyClass() {
        @Override
        public void start(Object parent) {
        }

        @Override
        public void end(Object parent) {
        }
    };
}
```

config.xml

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="ASCII" />
    <property name="severity" value="error" />
    <property name="fileExtensions" value="java" />

    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="scope" value="public" />
            <property name="allowedAnnotations" value="BeforeClass, Before" />
            <property name="allowMissingPropertyJavadoc" value="true" />
            <property name="validateThrows" value="true" />
            <property name="ignoreMethodNamesRegex" value="^(staticSetUp)$" />
        </module>
    </module>
</module>
```

output from: `java -jar checkstyle-6.14.1-all.jar -c config.xml TestClass.java`

```
Starting audit...
Audit done.
```

Expected output:

```
Starting audit...
[ERROR] TestClass.java:6:9: Missing a Javadoc comment. [JavadocMethod]
[ERROR] TestClass.java:10:9: Missing a Javadoc comment. [JavadocMethod]
Audit done.
Checkstyle ends with 2 errors.
```

JavadocMethod should validate all methods, including methods of anonymous classes.
##

**Comment by rnveach** [comment]
Most likely this is because of issues we have had with Scope enumeration and should switch to AccessModifier here like we did for ParameterNameCheck.

**Comment by DXTkastb** [comment]
I'm on it.

**Comment by DXTkastb** [comment]
> Most likely this is because of issues we have had with `Scope` enumeration and should switch to `AccessModifier` here like we did for `ParameterNameCheck`.

@rnveach , I missed your comment on this issue before working on this. I have another small patch which might fix this issue:

if (ast.getType()== TokenTypes.METHOD_DEF && as.getType() == TokenTypes.LITERAL_NEW)

this condition checks if ast node is method and also method ast node parent's parent is 'new' type node
![gghghghghg](https://user-images.githubusercontent.com/38028330/79699846-9a6ddb80-82af-11ea-8bd4-2a7eefaf3557.png)

if its true we'll skip condition `nodeScope.isIn(scope);` which actually hinders the Check to verify anonymous class methods[ scope=private, nodeScope=anoninner ].

I want your feedback on this, should I do it this way or change the Check entirely to verify using AccessModifier.

**Comment by DXTkastb** [comment]
Also,
I some pitest related issue:

https://dxtkastb.github.io/javadocpit/202004200142/com.puppycrawl.tools.checkstyle.checks.javadoc/JavadocMethodCheck.java.html#org.pitest.mutationtest.report.html.SourceFile@36782bf3_447

I removed condition: && nodeScope.isIn(scope); 
and pitest goes fine with it. I think tests must be added for this?

https://dxtkastb.github.io/javadocpit/202004200149/com.puppycrawl.tools.checkstyle.checks.javadoc/JavadocMethodCheck.java.html#org.pitest.mutationtest.report.html.SourceFile@30a17dbd_447

**Comment by kshivam953** [comment]
@romani @rnveach I wanna work on this issue. Please assign me this since I'm a new to this repo so, please help me through. Please help me finding where to start first.

**Comment by shubh220922** [comment]
@romani @nrmancuso  
What's the status here? Last comment was in april 2020.
No fixes made for this issue so far. Can I work on this?

**Comment by nrmancuso** [comment]
@shubh220922 if you are waiting on mentors on all PRs, you can work on this, but it might be a bit before we can get to review.

**Comment by romani** [comment]
@rnveach , can you review this issue again.
Documentation on inner classes are never goes to html report of javadoc tool. So in this case why we need violations?
I have never seen people to care about javadoc on anonymous classes.
So no violation is good actually :)

**Comment by rnveach** [comment]
This was when we still had `scope` as a property which supported anonymous classes. I don't see anywhere in the documentation it says it won't validate methods.

I care about javadocs on all methods as it was my mandate in my project to document all methods that we defined. I don't care about the HTML report really, but want documentation on methods we create.

**Comment by romani** [comment]
I think we need special property for this, not enabled by default. A lot of people will not like this.

**Comment by shubh220922** [comment]
I'm working on this.

**Comment by shubh220922** [comment]
I'm working on this issue.

**Comment by shubh220922** [comment]
I'm on it.

**Comment by AayushSaini101** [comment]
@romani I am on it

**Comment by romani** [comment]
@rnveach, to get expected result like you requested you can use different Check:

```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
      <module name="MissingJavadocMethod">
        <property name="scope" value="anoninner"/>
        <property name="allowedAnnotations" value=""/>
      </module>
    </module>
</module>

$ cat Test.java 
public class TestClass {
    public static final MyClass myField = new MyClass() {
        @Override
        public void start(Object parent) {
        }

        @Override
        public void end(Object parent) {
        }
    };
}

$ java -jar checkstyle-10.12.4-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:3:9: Missing a Javadoc comment. [MissingJavadocMethod]
[ERROR] /var/tmp/Test.java:7:9: Missing a Javadoc comment. [MissingJavadocMethod]
Audit done.
Checkstyle ends with 2 errors.
```

Do you need just detection of comment or full validation of javadoc ?

**Comment by romani** [comment]
we can reopen https://github.com/checkstyle/checkstyle/pull/13968/files where we update JavadocMethod to validate methods in anonymous classes if comments over methods are javadoc `/**`.
if no javadoc comment on method of anon class, we do not do violation by JavadocMethod.

I like this approach. if you made javadoc comment `/**` follow all rules of javadoc, if you do not want follow all rules of javadoc use `/*`, if MissingJavadocMethod is not configured for anonymous we do not demand javadoc.

**Comment by rnveach** [comment]
If my issue is resolved, then I am fine closing this.
